elrond-wasm/src/storage/mappers/user_mapper.rs

```diff
@@ -0,0 +1,176 @@
+use super::StorageMapper;
+use crate::abi::{TypeAbi, TypeName};
+use crate::api::{EndpointFinishApi, ErrorApi, StorageReadApi, StorageWriteApi};
+use crate::io::EndpointResult;
+use crate::storage::{storage_get, storage_set};
+use crate::types::{Address, BoxedBytes, MultiResultVec};
+use alloc::vec::Vec;
+
+const ADDRESS_TO_ID_SUFFIX: &[u8] = b"_address_to_id";
+const ID_TO_ADDRESS_SUFFIX: &[u8] = b"_id_to_address";
+const COUNT_SUFFIX: &[u8] = b"_count";
+
+/// Very widely used mapper, that manages the users of a smart contract.
+/// It holds a bi-directional map, from addresses to ids and viceversa.
+/// This is so we can easily iterate over all users, using their ids.
+/// Also holds the user count in sync. This is also necessary for iteration.
+///
+/// This particular implementation of a user mapper doesn't contain any additional
+/// user data other than address/id.
+///
+/// It also doesn't allow removing users. Once in, their ids are reserved forever.
+pub struct UserMapper<SA>
+where
+	SA: StorageReadApi + StorageWriteApi + ErrorApi + Clone + 'static,
+{
+	api: SA,
+	main_key: BoxedBytes,
+}
+
+impl<SA> StorageMapper<SA> for UserMapper<SA>
+where
+	SA: StorageReadApi + StorageWriteApi + ErrorApi + Clone + 'static,
+{
+	fn new(api: SA, main_key: BoxedBytes) -> Self {
+		UserMapper { api, main_key }
+	}
+}
+
+impl<SA> UserMapper<SA>
+where
+	SA: StorageReadApi + StorageWriteApi + ErrorApi + Clone + 'static,
+{
+	fn get_user_id_key(&self, address: &Address) -> BoxedBytes {
+		// TODO: build this more elegantly from elrond-codec
+		BoxedBytes::from_concat(&[
+			self.main_key.as_slice(),
+			&ADDRESS_TO_ID_SUFFIX[..],
+			address.as_bytes(),
+		])
+	}
+
+	/// Yields the user id for a given address.
+	/// Will return 0 if the address is not known to the contract.
+	pub fn get_user_id(&self, address: &Address) -> usize {
+		storage_get(self.api.clone(), self.get_user_id_key(address).as_slice())
+	}
+
+	fn set_user_id(&self, address: &Address, id: usize) {
+		storage_set(
+			self.api.clone(),
+			self.get_user_id_key(address).as_slice(),
+			&id,
+		);
+	}
+
+	fn get_user_address_key(&self, id: usize) -> BoxedBytes {
+		// TODO: build this more elegantly from elrond-codec
+		let id_bytes = (id as u32).to_be_bytes();
+		BoxedBytes::from_concat(&[
+			self.main_key.as_slice(),
+			&ID_TO_ADDRESS_SUFFIX[..],
+			&id_bytes[..],
+		])
+	}
+
+	/// Yields the user address for a given id, if the id is valid.
+	pub fn get_user_address(&self, id: usize) -> Option<Address> {
+		let key = self.get_user_address_key(id);
+		// TODO: optimize, storage_load_len is currently called twice
+		if self.api.storage_load_len(key.as_slice()) > 0 {
+			Some(storage_get(self.api.clone(), key.as_slice()))
+		} else {
+			None
+		}
+	}
```

Comment:
Maybe add a check for id = 0, but it isn't really necessary.

Reply:
Should return None for 0, but I didn't get to add tests for that yet. Work in progress.